GDR Cannot register memory region

### System information
- **Have I written custom code (as opposed to using a stock example script provided in TensorFlow)**: NO
- **OS Platform and Distribution (e.g., Linux Ubuntu 16.04)**:  Linux Ubuntu 16.04
- **Mobile device (e.g. iPhone 8, Pixel 2, Samsung Galaxy) if the issue happens on mobile device**: no
- **TensorFlow installed from (source or binary)**: source
- **TensorFlow version (use command below)**:1.10
- **Python version**: 2.7 
- **Bazel version (if compiling from source)**: 0.16.1
- **GCC/Compiler version (if compiling from source)**: 5.4
- **CUDA/cuDNN version**: 9.0
- **GPU model and memory**: K40c 12G

## Describe the problem
I tried to use GDR, but it seems no performance improvement compared to native RDMA, and logs are the following:

```
$ python benchmarks/scripts/tf_cnn_benchmarks/tf_cnn_benchmarks.py --local_parameter_device=cpu --num_gpus=4 --batch_size=128 --num_epochs=10 --model=alexnet --variable_update=distributed_replicated --data_dir=/home/shuai/imagenet-data --all_reduce_spec=pscpu --job_name=worker --ps_hosts=10.10.10.6:2222 --worker_hosts=10.10.10.5:3333 --task_index=0 --server_protocol=grpc+gdr
2018-09-05 15:05:58.062100: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1405] Found device 0 with properties:
name: Tesla K40c major: 3 minor: 5 memoryClockRate(GHz): 0.745
pciBusID: 0000:02:00.0
totalMemory: 11.92GiB freeMemory: 11.84GiB
2018-09-05 15:05:58.211502: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1405] Found device 1 with properties:
name: Tesla K40c major: 3 minor: 5 memoryClockRate(GHz): 0.745
pciBusID: 0000:03:00.0
totalMemory: 11.92GiB freeMemory: 11.84GiB
2018-09-05 15:05:58.374791: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1405] Found device 2 with properties:
name: Tesla K40c major: 3 minor: 5 memoryClockRate(GHz): 0.745
pciBusID: 0000:83:00.0
totalMemory: 11.92GiB freeMemory: 11.84GiB
2018-09-05 15:05:58.553152: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1405] Found device 3 with properties:
name: Tesla K40c major: 3 minor: 5 memoryClockRate(GHz): 0.745
pciBusID: 0000:84:00.0
totalMemory: 11.92GiB freeMemory: 11.84GiB
2018-09-05 15:05:58.553735: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1484] Adding visible gpu devices: 0, 1, 2, 3
2018-09-05 15:05:59.918282: I tensorflow/core/common_runtime/gpu/gpu_device.cc:965] Device interconnect StreamExecutor with strength 1 edge matrix:
2018-09-05 15:05:59.918335: I tensorflow/core/common_runtime/gpu/gpu_device.cc:971]      0 1 2 3
2018-09-05 15:05:59.918344: I tensorflow/core/common_runtime/gpu/gpu_device.cc:984] 0:   N Y N N
2018-09-05 15:05:59.918349: I tensorflow/core/common_runtime/gpu/gpu_device.cc:984] 1:   Y N N N
2018-09-05 15:05:59.918354: I tensorflow/core/common_runtime/gpu/gpu_device.cc:984] 2:   N N N Y
2018-09-05 15:05:59.918359: I tensorflow/core/common_runtime/gpu/gpu_device.cc:984] 3:   N N Y N
2018-09-05 15:05:59.919609: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1097] Created TensorFlow device (/job:worker/replica:0/task:0/device:GPU:0 with 11473 MB memory) -> physical GPU (device: 0, name: Tesla K40c, pci bus id: 0000:02:00.0, compute capability: 3.5)
2018-09-05 15:06:00.088804: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1097] Created TensorFlow device (/job:worker/replica:0/task:0/device:GPU:1 with 11473 MB memory) -> physical GPU (device: 1, name: Tesla K40c, pci bus id: 0000:03:00.0, compute capability: 3.5)
2018-09-05 15:06:00.288448: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1097] Created TensorFlow device (/job:worker/replica:0/task:0/device:GPU:2 with 11473 MB memory) -> physical GPU (device: 2, name: Tesla K40c, pci bus id: 0000:83:00.0, compute capability: 3.5)
2018-09-05 15:06:00.487520: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1097] Created TensorFlow device (/job:worker/replica:0/task:0/device:GPU:3 with 11473 MB memory) -> physical GPU (device: 3, name: Tesla K40c, pci bus id: 0000:84:00.0, compute capability: 3.5)
2018-09-05 15:06:00.688704: I tensorflow/core/distributed_runtime/rpc/grpc_channel.cc:215] Initialize GrpcChannelCache for job ps -> {0 -> 10.10.10.6:2222}
2018-09-05 15:06:00.688752: I tensorflow/core/distributed_runtime/rpc/grpc_channel.cc:215] Initialize GrpcChannelCache for job worker -> {0 -> localhost:3333}
2018-09-05 15:06:00.697485: I tensorflow/contrib/gdr/gdr_memory_manager.cc:254] RDMA server is listening on 10.10.10.5:3333
2018-09-05 15:06:00.697568: I tensorflow/contrib/gdr/gdr_memory_manager.cc:302] Instrumenting CPU allocator cuda_host_bfc
2018-09-05 15:06:00.697588: I tensorflow/contrib/gdr/gdr_memory_manager.cc:302] Instrumenting CPU allocator cpu_pool
2018-09-05 15:06:00.697607: I tensorflow/contrib/gdr/gdr_memory_manager.cc:302] Instrumenting CPU allocator cpu_rdma_bfc
2018-09-05 15:06:00.697830: I tensorflow/contrib/gdr/gdr_memory_manager.cc:95] NUMA node for device: mlx4_0 is 1
2018-09-05 15:06:00.734616: W tensorflow/contrib/gdr/gdr_memory_manager.cc:705] Cannot register memory region
2018-09-05 15:06:00.771893: W tensorflow/contrib/gdr/gdr_memory_manager.cc:705] Cannot register memory region
2018-09-05 15:06:00.771954: I tensorflow/contrib/gdr/gdr_memory_manager.cc:314] Instrumenting GPU allocator with bus_id 2
2018-09-05 15:06:00.772783: I tensorflow/core/distributed_runtime/rpc/grpc_server_lib.cc:375] Started server with target: grpc://localhost:3333
```
and here is GPU topology；
···
$ nvidia-smi topo -m
        GPU0    GPU1    GPU2    GPU3    mlx4_0  CPU Affinity
GPU0     X      PHB     SYS     SYS     SYS     0-7,16-23
GPU1    PHB      X      SYS     SYS     SYS     0-7,16-23
GPU2    SYS     SYS      X      PHB     PHB     8-15,24-31
GPU3    SYS     SYS     PHB      X      PHB     8-15,24-31
mlx4_0  SYS     SYS     PHB     PHB      X

Legend:

  X    = Self
  SYS  = Connection traversing PCIe as well as the SMP interconnect between NUMA nodes (e.g., QPI/UPI)
  NODE = Connection traversing PCIe as well as the interconnect between PCIe Host Bridges within a NUMA node
  PHB  = Connection traversing PCIe as well as a PCIe Host Bridge (typically the CPU)
  PXB  = Connection traversing multiple PCIe switches (without traversing the PCIe Host Bridge)
  PIX  = Connection traversing a single PCIe switch
  NV#  = Connection traversing a bonded set of # NVLinks
···

The logs said," cannot register memory region" in GPU 2 and GPU 3. Is it a bug?

Hi, @byronyi 

Sorry for late reply. I tried to replace the original code with the modified code in gdr_memory_manager.cc, based on tf r 1.10.1. But it failed. Then I downloaded the master source code and tried to install it, but it failed again with the following error messages:
```
ERROR: /home/shuai/tensorflow-master/tensorflow/contrib/gdr/BUILD:61:1: C++ compilati
tensorflow/contrib/gdr/gdr_worker.cc: In constructor 'tensorflow::GdrWorker::GdrWorke
tensorflow/contrib/gdr/gdr_worker.cc:46:45: error: no matching function for call to '
       recv_tensor_recent_request_ids_(100000) {}
                                             ^
In file included from ./tensorflow/contrib/gdr/gdr_worker.h:22:0,
                 from tensorflow/contrib/gdr/gdr_worker.cc:16:
./tensorflow/core/distributed_runtime/rpc/grpc_worker_service.h:36:3: note: candidate
   GrpcWorker(WorkerEnv* env, const ConfigProto& config);
   ^
./tensorflow/core/distributed_runtime/rpc/grpc_worker_service.h:36:3: note:   candida
Target //tensorflow/tools/pip_package:build_pip_package failed to build
```

--- a/tensorflow/contrib/gdr/gdr_server_lib.cc
+++ b/tensorflow/contrib/gdr/gdr_server_lib.cc
@@ -52,9 +52,10 @@ Status GdrServer::Init() {
       [this](const WorkerEnv* env) {
         return new GdrRendezvousMgr(env, remote_memory_manager_.get());
       };
-  WorkerCreationFunction worker_func = [this](WorkerEnv* env) {
+  WorkerCreationFunction worker_func = [this](WorkerEnv* env,
+                                              const ConfigProto& config) {
     return std::unique_ptr<GdrWorker>(
-        new GdrWorker(env, remote_memory_manager_.get()));
+        new GdrWorker(env, config, remote_memory_manager_.get()));
   };
 
   TF_RETURN_IF_ERROR(remote_memory_manager_->Init());

--- a/tensorflow/contrib/gdr/gdr_worker.cc
+++ b/tensorflow/contrib/gdr/gdr_worker.cc
@@ -39,9 +39,9 @@ limitations under the License.
 
 namespace tensorflow {
 
-GdrWorker::GdrWorker(WorkerEnv* worker_env,
+GdrWorker::GdrWorker(WorkerEnv* worker_env, const ConfigProto& config,
                      RemoteMemoryManager* remote_memory_manager)
-    : GrpcWorker(worker_env),
+    : GrpcWorker(worker_env, config),
       remote_memory_manager_(remote_memory_manager),
       recv_tensor_recent_request_ids_(100000) {}
 

--- a/tensorflow/contrib/gdr/gdr_worker.h
+++ b/tensorflow/contrib/gdr/gdr_worker.h
@@ -25,7 +25,8 @@ namespace tensorflow {
 
 class GdrWorker : public GrpcWorker {
  public:
-  GdrWorker(WorkerEnv* env, RemoteMemoryManager* remote_memory_manager);
+  GdrWorker(WorkerEnv* env, const ConfigProto& config,
+            RemoteMemoryManager* remote_memory_manager);
 
   // Serve the RecvTensorRequest but omit the tensor content and transmit it
   // out-of-band using GPU Direct RDMA whenever possible.

--- a/tensorflow/core/distributed_runtime/rpc/grpc_server_lib.cc
+++ b/tensorflow/core/distributed_runtime/rpc/grpc_server_lib.cc
@@ -194,7 +194,7 @@ Status GrpcServer::Init(
   MaybeMutateBuilder(&builder);
   master_impl_ = CreateMaster(&master_env_);
   master_service_ = NewGrpcMasterService(master_impl_.get(), config, &builder);
-  worker_impl_ = worker_func ? worker_func(&worker_env_)
+  worker_impl_ = worker_func ? worker_func(&worker_env_, config)
                              : NewGrpcWorker(&worker_env_, config);
   worker_service_ =
       NewGrpcWorkerService(worker_impl_.get(), &builder).release();

--- a/tensorflow/core/distributed_runtime/rpc/grpc_server_lib.h
+++ b/tensorflow/core/distributed_runtime/rpc/grpc_server_lib.h
@@ -53,7 +53,8 @@ typedef std::function<void(const WorkerEnv*, ::grpc::ServerBuilder*)>
     ServiceInitFunction;
 
 // function that creates a grpc based worker implementation.
-typedef std::function<std::unique_ptr<GrpcWorker>(WorkerEnv*)>
+typedef std::function<std::unique_ptr<GrpcWorker>(WorkerEnv*,
+                                                  const ConfigProto& config)>
     WorkerCreationFunction;
 
 class GrpcServer : public ServerInterface {

